src/main/java/io/camunda/zeebe/bpmnassert/extensions/ZeebeAssertionsExtension.java

```diff
@@ -1,77 +1,98 @@
+    final ZeebeClient client = (ZeebeClient) clientContent;
+    client.close();
+
+    final Object engineContent = getStore(extensionContext).get(KEY_ZEEBE_ENGINE);
+    final InMemoryEngine engine = (InMemoryEngine) engineContent;
+    engine.stop();
   }
 
-  private void storeRecordStreamSource(final ExtensionContext extensionContext)
-      throws IllegalAccessException {
-    final Field recordStreamField = getRecordStreamField(extensionContext);
-    ReflectionUtils.makeAccessible(recordStreamField);
-    final RecordStreamSource recordStreamSource =
-        (RecordStreamSource) recordStreamField.get(extensionContext.getRequiredTestInstance());
-    RecordStreamSourceStore.init(recordStreamSource);
+  @Override
+  public void testFailed(final ExtensionContext extensionContext, final Throwable cause) {
+    final Object engineContent = getStore(extensionContext).get(KEY_ZEEBE_ENGINE);
+    final InMemoryEngine engine = (InMemoryEngine) engineContent;
+
+    System.out.println("===== Test failed! Printing records from the stream:");
+    engine.getRecordStream().print(true);
   }
 
-  private Field getRecordStreamField(final ExtensionContext extensionContext) {
-    final Optional<Field> recordStreamOptional =
-        Arrays.stream(extensionContext.getRequiredTestClass().getDeclaredFields())
-            .filter(field -> field.getType() == RecordStreamSource.class)
-            .findFirst();
-    return recordStreamOptional.orElseThrow(
-        () ->
-            new IllegalStateException(
-                "Expected a field of type RecordStreamSource to be declared in the test class, "
-                    + "but none has been found. Please make sure a field of type RecordStreamSource"
-                    + " has been declared in the test class."));
+  private void injectFields(final ExtensionContext extensionContext, final Object... objects) {
+    final Class<?> requiredTestClass = extensionContext.getRequiredTestClass();
+    final Field[] declaredFields = requiredTestClass.getDeclaredFields();
+    for (Object object : objects) {
+      final Optional<Field> field = getField(declaredFields, object);
+      field.ifPresent(value -> injectField(extensionContext, value, object));
+    }
   }
 
-  private void storeZeebeClient(final ExtensionContext extensionContext)
-      throws IllegalAccessException {
-    final Field zeebeClientField = getZeebeClientField(extensionContext);
-    ReflectionUtils.makeAccessible(zeebeClientField);
-    final ZeebeClient zeebeClient =
-        (ZeebeClient) zeebeClientField.get(extensionContext.getRequiredTestInstance());
-    getStore(extensionContext).put(KEY_ZEEBE_CLIENT, zeebeClient);
+  private Optional<Field> getField(final Field[] declaredFields, final Object object) {
+    final List<Field> fields =
+        Arrays.stream(declaredFields)
+            .filter(field -> field.getType().isInstance(object))
+            .collect(Collectors.toList());
+
+    if (fields.size() > 1) {
```

Comment:
:question: Why is this important? We could inject it in all fields anyway

Reply:
We could. I just thought it might be confusing from a user perspective if he defines two clients that they will both be the same client. And I don't think we should manage multiple clients in our extension.

I don't see any reason why anyone would need the fields twice so I decided prevent people from doing this and causing unexpected behaviours.